src/bin/ambit/cmd.rs

```diff
@@ -52,24 +53,122 @@ fn is_symlinked(link_name: &Path, target: &Path) -> bool {
+                        }
+                    }
+                }
+                Ok(matches)
+            }
+            // The only valid path at the start is the starting path.
+            let mut valid_paths: Vec<PathBuf> = vec![start_path.clone()];
+            let components: Vec<String> = Path::new(&entry)
+                .components()
+                .map(|comp| comp.as_os_str().to_string_lossy().to_string())
+                .collect();
+            for (i, component) in components.iter().enumerate() {
+                let mut new_valid_paths: Vec<PathBuf> = Vec::new();
+                let expected_path_kind = if i < components.len() - 1 {
+                    // There are still more components to go, expect a directory.
+                    AmbitPathKind::Directory
+                } else {
+                    // No more components, expect a file.
+                    AmbitPathKind::File
+                };
+                let pattern = Pattern::compile(
+                    &component,
+                    MatchOptions::WILDCARDS | MatchOptions::UNKNOWN_CHARS,
+                );
+                for path in &valid_paths {
+                    new_valid_paths.append(&mut get_matches(
+                        &path,
+                        &expected_path_kind,
+                        |name: &str| -> bool { pattern.matches(name) },
+                    )?);
+                }
+                valid_paths = new_valid_paths;
+            }
+            // Strip prefix from all paths.
+            valid_paths = valid_paths
+                .iter()
+                .map(|path| -> AmbitResult<PathBuf> {
+                    Ok(path.strip_prefix(&start_path)?.to_path_buf())
+                })
+                .collect::<AmbitResult<Vec<PathBuf>>>()?;
+            paths.append(&mut valid_paths);
+        } else {
+            paths.push(PathBuf::from(&entry));
+        }
+    }
+    Ok(paths)
+}
+
 // Return iterator over path pairs in the form of `(repo_file, host_file)` from given entry.
-fn get_ambit_paths_from_entry<'a>(
-    entry: &'a Entry,
-) -> Box<dyn Iterator<Item = (AmbitPath, AmbitPath)> + 'a> {
-    Box::new(
-        entry
-            .left
-            .into_iter()
-            // If entry.right is None, entry.left is considered to be both repo and host path.
-            .zip(entry.right.as_ref().unwrap_or(&entry.left).into_iter())
-            .map(|(repo_path, host_path)| {
-                // Wrap the given paths as AmbitPath.
-                (
-                    AmbitPath::new(AMBIT_PATHS.repo.path.join(repo_path), AmbitPathKind::File),
-                    AmbitPath::new(AMBIT_PATHS.home.path.join(host_path), AmbitPathKind::File),
-                )
-            }),
-    )
+fn get_ambit_paths_from_entry(entry: &Entry) -> AmbitResult<Vec<(AmbitPath, AmbitPath)>> {
+    let (left_paths, right_paths): (Vec<PathBuf>, Option<Vec<PathBuf>>) =
+        if let Some(entry_right) = &entry.right {
```

Comment:
This `if` statement can be removed, to obtain something similar to:
```
let left_entry_start = if entry.right.is_some() { <home> } else { <repo> };
(
  <process_left_entry>,
  if let Some(entry_right) = &entry.right {
    Some(<process_right_entry>)
  } else {
    None
  }
)
```
The variables `left_paths` and `right_paths` should also not be assigned in a tuple.